tests/bdd/test_volume_topology.py

```diff
@@ -0,0 +1,536 @@
+
+
+@when("a user attempts to increase the number of volume replicas")
+def a_user_attempts_to_increase_the_number_of_volume_replicas(replica_ctx):
+    """a user attempts to increase the number of volume replicas."""
+    volumes_api = common.get_volumes_api()
+    volume = volumes_api.get_volume(VOLUME_UUID)
+    num_replicas = volume.spec.num_replicas
+    try:
+        volume = volumes_api.put_volume_replica_count(VOLUME_UUID, num_replicas + 1)
+        replica_ctx[REPLICA_CONTEXT_KEY] = volume.spec.num_replicas
+    except Exception as e:
+        replica_ctx[REPLICA_ERROR] = e
+
+
+@when("the number of suitable pools is less than the number of desired volume replicas")
+def the_number_of_suitable_pools_is_less_than_the_number_of_desired_volume_replicas(
+    create_request,
+):
+    """the number of suitable pools is less than the number of desired volume replicas."""
+    num_volume_replicas = create_request[CREATE_REQUEST_KEY]["replicas"]
+    no_of_pools = no_of_suitable_pools(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ]
+    )
+    assert num_volume_replicas > no_of_pools
+
+
+@when(
+    "the number of volume replicas is less than or equal to the number of suitable pools"
+)
+def the_number_of_volume_replicas_is_less_than_or_equal_to_the_number_of_suitable_pools(
+    create_request,
+):
+    """the number of volume replicas is less than or equal to the number of suitable pools."""
+    num_volume_replicas = create_request[CREATE_REQUEST_KEY]["replicas"]
+    # Here none denotes that pool topology is none and thus all pools are eligible
+    no_of_pools = no_of_suitable_pools(None)
+    assert num_volume_replicas <= no_of_pools
+
+
+@then("an additional replica should be added to the volume")
+def an_additional_replica_should_be_added_to_the_volume(replica_ctx):
+    """an additional replica should be added to the volume."""
+    volume = common.get_volumes_api().get_volume(VOLUME_UUID)
+    assert hasattr(volume.state, "target")
+    nexus = volume.state.target
+    assert replica_ctx[REPLICA_CONTEXT_KEY] == len(nexus["children"])
+    assert REPLICA_ERROR not in replica_ctx
+
+
+@then("pool labels must contain all the volume request topology labels")
+def pool_labels_must_contain_all_the_volume_request_topology_labels(create_request):
+    """pool labels must contain all the volume request topology labels."""
+    assert common_labels(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ],
+        common.get_pools_api().get_pool(POOL_1_UUID),
+    ) or common_labels(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ],
+        common.get_pools_api().get_pool(POOL_2_UUID),
+    ) == len(
+        create_request[CREATE_REQUEST_KEY]["topology"]["pool_topology"]["labelled"][
+            "inclusion"
+        ]
+    )
```

Comment:
I think you might want to check all these complex assertions for parenthesis. At the moment it is of the form `a or b == len(c)` but I think it needs to be `(a or b) == len(c)`. The order of precedence matters here.

Additionally, in this statement I think the `or` is meant to be an `and` because we want to check that both pools contain the correct number of labels.

Reply:
Ah ok. Then we just need the parenthesis.